api/src/main/java/io/grpc/NameResolverRegistry.java

```diff
@@ -149,23 +151,17 @@ public static synchronized NameResolverRegistry getDefaultRegistry() {
     @Override
     @Nullable
     public NameResolver newNameResolver(URI targetUri, NameResolver.Args args) {
-      List<NameResolverProvider> providers = providers();
-      for (NameResolverProvider provider : providers) {
-        NameResolver resolver = provider.newNameResolver(targetUri, args);
-        if (resolver != null) {
-          return resolver;
-        }
-      }
-      return null;
+      NameResolverProvider provider = providers().get(targetUri.getScheme());
+      return provider == null ? null : provider.newNameResolver(targetUri, args);
     }
 
     @Override
     public String getDefaultScheme() {
-      List<NameResolverProvider> providers = providers();
+      LinkedHashMap<String, NameResolverProvider> providers = providers();
       if (providers.isEmpty()) {
         return "unknown";
       }
-      return providers.get(0).getDefaultScheme();
+      return providers.entrySet().iterator().next().getKey();
```

Comment:
I don't think this does what you are hoping. This does not get the first entry with the highest priority, as items are inserted into the map in `allProviders`-order.

It seems easiest to just compute this value within `refreshProviders()`, either manually within the loop or using something like `Collections.max()` (although max() doesn't guarantee any particular behavior when multiple entries are maximal).

Reply:
yea we probably should return the highest priority provider scheme as the default scheme. Then I guess we do not need linkedhashmap anymore just a normal immutable hashmap. `allProviders` will guarantee we get the consistent default scheme.